scripts/dats_jsonld_annotator/annotator.py

```diff
@@ -0,0 +1,323 @@
+import json
+import logging
+import pathlib as pal
+import sys
+import time
+
+import jsonschema as jss
+
+
+CONTEXT_DIR = (pal.Path(__file__).parent / "context" / "sdo").resolve()
+SCHEMA_DIR = (pal.Path(__file__).parent / "schema").resolve()
+logger = logging.getLogger("DATS annotator")
+
+
+def find_schema(parent_schema, term, json_object):
+    """
+    This function finds the appropriate JSON schema for a term in a DATS Json object. To do
+    so, it relies on the JSON schema of the supplied JSON object.
+
+    :param parent_schema: The DATS JSON schema that corresponds to the json_object parameter
+    :param term: the string term for which the schema shall be found
+    :param json_object:  the JSON object that contains the term and to which the parent_schema corresponds
+    :return: the DATS JSON schema for the term
+    """
+    ps = parent_schema["properties"]
+    if term not in ps.keys():
+        # TODO: Decide whether to raise an Exception here
+        # The only way this could happen on a valid dataset is if someone has specified additional properties
+        # We'd like to discourage this
+        logger.debug(
+            f'I cannot find {term = } in the parent schema: {parent_schema["id"]}.\n{json_object = }'
+        )
+        return None
+
+    # Set up the schema resolver
+    resolver = jss.RefResolver(base_uri=parent_schema["id"], referrer=None)
+
+    # See if "items" is in the schema entry
+    search_dict = ps[term]
+    if "items" in search_dict.keys():
+        search_dict = ps[term]["items"]
+
+    if "$ref" in search_dict.keys():
+        schema_name = search_dict["$ref"]
+    elif len(set(search_dict.keys()).intersection(["anyOf", "oneOf", "allOf"])) > 0:
+        # Get the correct jsonld relationship
+        schema_rel = list(
+            set(search_dict.keys()).intersection(["anyOf", "oneOf", "allOf"])
+        )[0]
+        possible_keys = []
+        for ref in search_dict[schema_rel]:
+            # Get the schema
+            ref_name = ref["$ref"]
+            _schema_uri, _schema = resolver.resolve(ref_name)
+            if _schema_uri == resolver.base_uri:
+                _schema = parent_schema
+            if jss.Draft4Validator(_schema).is_valid(json_object):
+                possible_keys.append(ref_name)
+        if len(possible_keys) > 1:
+            # TODO: decide if we let the user pick which option to go with
+            logger.debug(f"I got more than one option for {term}: {possible_keys}")
+        elif len(possible_keys) == 0:
+            # TODO: we might also want to raise here as this shouldn't happen
+            logger.warning(
+                f"I have no fitting schema for {json_object} {term} among {search_dict[schema_rel]}"
+            )
+            return None
+        # If anything fits, just pick the first one
+        # TODO: we may want to leave this up to the user here, particularly if the instances
+        #       map to different / meaningful things in SDO
```

Comment:
What do you have in mind here? It's pretty frequent that more than one schema option would be valid. So for now I logged this as "debug" and don't show by default. If we were to raise the log level, these messages would most likely overwhelm / dominate the terminal output. But I agree, this information is relevant.

Regarding the TODO: I think letting the user choose the "correct" schema sounds like a good strategy. But the reality is that the majority of the schema @types (i.e. the schema.org classes) are sdo:Thing. So in the final jsonld output it probably rarely really makes a difference if we choose one or the other from a list of valid schemata. Given how frequently such a choice would have to be made, this would make the annotation pretty laborious / slow. So we should probably discuss this together.